From 2a841d0d1c511b8df74afefa7b4fcdf43c72c80e Mon Sep 17 00:00:00 2001
From: Connor Abbott <cwabbott0@gmail.com>
Date: Thu, 21 May 2026 12:23:39 -0400
Subject: [PATCH] ir3: Consider speculability when lowering UBOs to consts

We may not be able to lower UBOs to consts if the descriptor set may be
partially bound in Vulkan. Consider speculability to fix this. Since
opt_preamble can correctly conditionalize these non-speculatable loads,
change the heuristic to let opt_preamble handle them at the cost of a
somewhat larger/slower preamble.

Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/42159>
(cherry picked from mesa main commit 2a841d0d1c511b8df74afefa7b4fcdf43c72c80e)
---
 src/compiler/nir/nir_intrinsics.py             |  4 ++--
 src/freedreno/ir3/ir3_nir_analyze_ubo_ranges.c | 16 ++++++++++++++--
 src/freedreno/ir3/ir3_nir_opt_preamble.c       | 13 +++++++------
 src/freedreno/ir3/ir3_shader.h                 |  1 +
 4 files changed, 24 insertions(+), 10 deletions(-)

diff --git a/src/compiler/nir/nir_intrinsics.py b/src/compiler/nir/nir_intrinsics.py
index 35af592b184..db479f43169 100644
--- a/src/compiler/nir/nir_intrinsics.py
+++ b/src/compiler/nir/nir_intrinsics.py
@@ -1707,12 +1707,12 @@ load("const_ir3", [1], indices=[BASE], flags=[CAN_ELIMINATE, CAN_REORDER])
 # IR3-specific intrinsic for ldc.k. Copies UBO to constant file.
 # base is the const file base in components, range is the amount to copy in
 # vec4's.
-intrinsic("copy_ubo_to_uniform_ir3", [1, 1], indices=[BASE, RANGE])
+intrinsic("copy_ubo_to_uniform_ir3", [1, 1], indices=[ACCESS, BASE, RANGE])
 
 # IR3-specific intrinsic for ldg.k.
 # base is an offset to apply to the address in bytes, range_base is the
 # const file base in components, range is the amount to copy in vec4's.
-intrinsic("copy_global_to_uniform_ir3", [1], indices=[BASE, RANGE_BASE, RANGE])
+intrinsic("copy_global_to_uniform_ir3", [1], indices=[ACCESS, BASE, RANGE_BASE, RANGE])
 
 # IR3-specific intrinsic for stsc. Loads from push consts to constant file
 # Should be used in the shader preamble.
diff --git a/src/freedreno/ir3/ir3_nir_analyze_ubo_ranges.c b/src/freedreno/ir3/ir3_nir_analyze_ubo_ranges.c
index 6e7e332331c..0c3570dbf82 100644
--- a/src/freedreno/ir3/ir3_nir_analyze_ubo_ranges.c
+++ b/src/freedreno/ir3/ir3_nir_analyze_ubo_ranges.c
@@ -162,6 +162,8 @@ gather_ubo_ranges(nir_shader *nir, nir_intrinsic_instr *instr,
 
       plan_r->start = r.start;
       plan_r->end = r.end;
+      plan_r->can_speculate &=
+         !!(nir_intrinsic_access(instr) & ACCESS_CAN_SPECULATE);
       *upload_remaining -= added;
 
       merge_neighbors(state, i);
@@ -179,6 +181,7 @@ gather_ubo_ranges(nir_shader *nir, nir_intrinsic_instr *instr,
    plan_r->ubo = ubo;
    plan_r->start = r.start;
    plan_r->end = r.end;
+   plan_r->can_speculate = nir_intrinsic_access(instr) & ACCESS_CAN_SPECULATE;
    *upload_remaining -= added;
 }
 
@@ -415,7 +418,8 @@ copy_global_to_uniform(nir_shader *nir, struct ir3_ubo_analysis_state *state)
       for (unsigned offset = 0; offset < size; offset += 256 * 16) {
          unsigned const_offset = range->offset / 4 + offset / 4;
          nir_copy_global_to_uniform_ir3(
-            b, base, .base = start + offset, .range_base = const_offset,
+            b, base, .access = range->can_speculate ? ACCESS_CAN_SPECULATE : 0,
+            .base = start + offset, .range_base = const_offset,
             .range = MIN2(256, (size - offset) / 16));
       }
    }
@@ -439,8 +443,11 @@ copy_ubo_to_uniform(nir_shader *nir, const struct ir3_const_state *const_state)
       const struct ir3_ubo_range *range = &state->range[i];
 
       nir_def *ubo = nir_imm_int(b, range->ubo.block);
+      enum gl_access_qualifier access =
+         range->can_speculate ? ACCESS_CAN_SPECULATE : 0;
       if (range->ubo.bindless) {
          ubo = nir_bindless_resource_ir3(b, 32, ubo,
+                                         .access = access,
                                          .desc_set = range->ubo.bindless_base);
       }
 
@@ -451,6 +458,7 @@ copy_ubo_to_uniform(nir_shader *nir, const struct ir3_const_state *const_state)
       for (unsigned offset = 0; offset < size; offset += 256) {
          nir_copy_ubo_to_uniform_ir3(b, ubo, nir_imm_int(b, range->start / 16 +
                                                          offset),
+                                     .access = access,
                                      .base = range->offset / 4 + offset * 4,
                                      .range = MIN2(size - offset, 256));
       }
@@ -470,7 +478,10 @@ instr_is_load_ubo(nir_instr *instr)
    /* nir_lower_ubo_vec4 happens after this pass. */
    assert(op != nir_intrinsic_load_ubo_vec4);
 
-   return op == nir_intrinsic_load_ubo;
+   if (op != nir_intrinsic_load_ubo)
+      return false;
+
+   return ir3_nir_is_prefetchable(nir_instr_as_intrinsic(instr));
 }
 
 bool
@@ -831,6 +842,7 @@ ir3_nir_lower_load_const_instr(nir_builder *b, nir_instr *in_instr, void *data)
 
    nir_def *result =
       nir_load_ubo(b, num_components, bit_size, index, offset,
+                   .access = ACCESS_CAN_SPECULATE,
                    .align_mul = nir_intrinsic_align_mul(instr),
                    .align_offset = nir_intrinsic_align_offset(instr),
                    .range_base = base, .range = nir_intrinsic_range(instr));
diff --git a/src/freedreno/ir3/ir3_nir_opt_preamble.c b/src/freedreno/ir3/ir3_nir_opt_preamble.c
index 1e90d1dd3b9..53bce135de5 100644
--- a/src/freedreno/ir3/ir3_nir_opt_preamble.c
+++ b/src/freedreno/ir3/ir3_nir_opt_preamble.c
@@ -160,11 +160,11 @@ instr_cost(nir_instr *instr, const void *data)
       nir_intrinsic_instr *intrin = nir_instr_as_intrinsic(instr);
       switch (intrin->intrinsic) {
       case nir_intrinsic_load_ubo: {
-         /* If the UBO and offset are constant, then UBO lowering should do a
-          * better job trying to lower this, and opt_preamble shouldn't try to
-          * duplicate it. However if it has a non-constant offset then we can
-          * avoid setting up a0.x etc. in the main shader and potentially have
-          * to push less.
+         /* If the UBO and offset are constant and it is speculatable, then UBO
+          * lowering should do a better job trying to lower this, and
+          * opt_preamble shouldn't try to duplicate it. However if it has a
+          * non-constant offset then we can avoid setting up a0.x etc. in the
+          * main shader and potentially have to push less.
           */
          bool const_ubo = nir_src_is_const(intrin->src[0]);
          if (!const_ubo) {
@@ -173,7 +173,8 @@ instr_cost(nir_instr *instr, const void *data)
                const_ubo = nir_src_is_const(rsrc->src[0]);
          }
 
-         if (const_ubo && nir_src_is_const(intrin->src[1]))
+         if (const_ubo && nir_src_is_const(intrin->src[1]) &&
+             ir3_nir_is_prefetchable(intrin))
             return 0;
 
          /* TODO: get actual numbers for ldc */
diff --git a/src/freedreno/ir3/ir3_shader.h b/src/freedreno/ir3/ir3_shader.h
index 5c3537f2df3..3e81f810932 100644
--- a/src/freedreno/ir3/ir3_shader.h
+++ b/src/freedreno/ir3/ir3_shader.h
@@ -165,6 +165,7 @@ struct ir3_ubo_range {
    struct ir3_ubo_info ubo;
    uint32_t offset;     /* start offset to push in the const register file */
    uint32_t start, end; /* range of block that's actually used */
+   bool can_speculate;
 };
 
 struct ir3_ubo_analysis_state {
-- 
2.55.0

